Refact stock entry purchase

lomamech opened the pull request:
refactor(Improvements Stock Entry) 
- Improve modal header and give text
- Use form-control-static in non editable mode
- Appear all Validation messages
- Turn off all validation until the submit method called
- Add the validation for excessive quatity
- Remove the untranslate label by aggregationHideLabel

closes #2015

lomamech committed:
refactor(Improvements Stock Entry)
- Improve modal header and give text
- Use form-control-static in non editable mode
- Appear all Validation messages
- Turn off all validation until the submit method called

lomamech committed:
refactor(Stock Entry)
- Add the validation for excessive quatity
- Remove the untranslate label by aggregationHideLabel

jniles commented on client/src/i18n/en/form.json:
```diff
@@ -90,6 +91,7 @@
             "MISSING_DEBTOR_ID": "Identifying missing Debtor",
             "RECORD_ERROR": "Please resolve any errors on the form (marked in red).",
             "REQUIRED": "This value is required.",
+            "PLEASE_CHECK_EXPIRY_DATE": "Please check the expiry date",
```
This should say "Please check the expiration date."

jniles commented on client/src/modules/stock/entry/modals/lots.modal.html:
```diff
@@ -4,17 +4,23 @@
+    <ol class="headercrumb">
+      <li class="static" translate>TREE.STOCK</li>
+      <li class="static" translate>STOCK.ENTRY</li>
+      <li class="title">{{ $ctrl.inventory.inventory.label }}</li>
```
:+1:

jniles commented on client/src/modules/stock/entry/modals/lots.modal.html:
```diff
@@ -69,6 +79,12 @@ <h3 style="margin: 0px;">{{ $ctrl.inventory.inventory.label }}</h3>
       <span translate>FORM.INFO.MAX_LOT_REACHED</span>
     </p>
+
+    <p ng-if="$ctrl.submitError" class="alert alert-danger" style="margin: 5px 0px;">
```
:+1:

jniles commented on client/src/modules/stock/entry/modals/lots.modal.html:
```diff
@@ -4,17 +4,23 @@
+
       <input 
-        ng-disabled="$ctrl.entryType === 'purchase' || $ctrl.entryType === 'transfer_reception'" 
+        ng-if="$ctrl.entryType !== 'purchase' && $ctrl.entryType !== 'transfer_reception'" 
```
Instead of repeating this check multiple times, a better way to express this would be in JS.  Something like this:
```js
// determine if the inventory quantity and cost should be editable or not.
$ctrl.hasEditableInventory = ($ctrl.entryType !== 'purchase' && $ctrl.entryType !== 'transfer_reception');
```
... later in the HTML...
```html
<div ng-if="$ctrl.hasEditableInventory">
   <!--- put all of the editable code here -->
</div>
<div ng-if="!$ctrl.hasEditableInventory">
  <!-- put all form-static code here -->
</div>
```

jniles commented on client/src/modules/stock/entry/modals/lots.modal.js:
```diff
@@ -34,13 +34,15 @@ function StockDefineLotsModalController(Instance, Notify, uiGridConstants, Data,
         displayName      : 'TABLE.COLUMNS.LOT',
         headerCellFilter : 'translate',
         aggregationType  : uiGridConstants.aggregationTypes.count,
+        aggregationHideLabel : true,
```
:dancing_men:

jniles commented on client/src/modules/stock/entry/modals/lots.modal.js:
```diff
@@ -34,13 +34,15 @@ function StockDefineLotsModalController(Instance, Notify, uiGridConstants, Data,
         displayName      : 'TABLE.COLUMNS.QUANTITY',
         headerCellFilter : 'translate',
         aggregationType  : uiGridConstants.aggregationTypes.sum,
+        aggregationHideLabel : true,
```
:dancing_women:

jniles commented on client/src/modules/stock/entry/modals/lots.modal.js:
```diff
@@ -124,6 +126,19 @@ function StockDefineLotsModalController(Instance, Notify, uiGridConstants, Data,
+
+    vm.submitError = vm.isExpiration;
+    if (vm.isExpiration) {
+      vm.errorText = 'FORM.ERRORS.PLEASE_CHECK_EXPIRY_DATE';
```
Since these errors are only related to Stock, it probably makes sense to put them in the stock en/fr JSON.  Something like `STOCK.ERRORS.PLEASE_CHECK_EXPIRY_DATE`.

jniles commented on client/src/modules/stock/entry/modals/lots.modal.js:
```diff
@@ -124,6 +126,19 @@ function StockDefineLotsModalController(Instance, Notify, uiGridConstants, Data,
     inventory.is_valid = (hasQuantity && hasLotLabel && hasExpiration);
 
+    // set the value if the date is not correct
+    vm.isExpiration = hasExpiration ? false : true;
```
Hmm... this binds a lot of variables to `vm` that you don't actually need.

How about your write a validation function for an inventory item (or lot, whichever makes sense)?  Something like this:
```js
function hasInvalidInventory(inventory) {
   // returns a string error like 'FORM.ERRORS.PLEASE_CHECK_EXPIRY_DATE' if there is an error.
  // otherwise, returns false.
}
```
You could then just call this function on every inventory item and it would return the appropriate message.
```js
var error = hasInvalidInventory(inventory);
if (error) {
  vm.submitErrror = true;
  vm.errorText = error;
}
```

This would reduce the logic needed to know the error type.  What do you think?

jniles reviewed (commented):
This is a step in the right direction.  The logic for error handling looks a bit scattered, I've suggested a function wrapper for it.  Let me know if you can find a way to clean this up.

lomamech committed:
refactor(Stock Entry)
- Improve translate key
- Use a functionhasEditableInventory for prevent multiple check lines
- Perf and add function haInvalidInventory for the validation

jniles reviewed (approved):
This is a good step in the right direction! :+1:

bors r+
